File: C++/Composite/TechCompany/employee.cc
#include <iostream>
#include <algorithm>
#include "employee.h"

Developer::Developer(const std::string& name, int employee_id, const std::string& title) : _name(name),
                                                                                           _employee_id(employee_id),
                                                                                           _title(title) {}

void Developer::show_details() {
    std::cout << "\t\tDEVELOPER ID: " << _employee_id << ", NAME: " << _name << ", TITLE: " << _title << "\n";
}



Salesperson::Salesperson(const std::string& name, int employee_id, const std::string& title) : _name(name),
                                                                                               _employee_id(employee_id),
                                                                                               _title(title) {}

void Salesperson::show_details() {
    std::cout << "\t\tSALESPERSON ID: " << _employee_id << ", NAME: " << _name << ", TITLE: " << _title << "\n";
}


Manager::Manager(const std::string& name, int employee_id, const std::string& title) : _name(name),
                                                                                       _employee_id(employee_id),
                                                                                       _title(title),
                                                                                       _employees() {}

void Manager::show_details() {
    std::cout << "\tMANAGER ID: " << _employee_id << ", NAME: " << _name << ", TITLE: " << _title << "\n";
    for (auto& employee : _employees) {
        employee->show_details();
    }
}

void Manager::add_employee(Employee* employee) {
    _employees.emplace(employee);
}

void Manager::remove_employee(Employee* employee) {
    auto it = std::find_if(_employees.begin(), _employees.end(),
                        [employee](auto& emp) {return emp.get() == employee;});
    _employees.erase(it);
}

CEO::CEO(const std::string& name, int employee_id, const std::string& title) : _name(name),
                                                                               _employee_id(employee_id),
                                                                               _title(title),
                                                                               _employees() {}

void CEO::add_employee(Employee* employee) {
    _employees.emplace(employee);
}

void CEO::remove_employee(Employee* employee) {
    auto it = std::find_if(_employees.begin(), _employees.end(),
                        [employee](auto& emp) {return emp.get() == employee;});
    _employees.erase(it);
}

void CEO::show_details() {
    std::cout << "CEO ID: " << _employee_id << ", NAME: " << _name << ", TITLE: " << _title << "\n";
    for (auto& employee : _employees) {
        employee->show_details();
    }
}

File: Java/Composite/TechCompany/src/com/ericgtkb/Employee.java
package com.ericgtkb;

import java.util.HashSet;
import java.util.Set;

// Component interface
// Sometimes add and remove is also put here
public interface Employee {
    void showDetails();
}

// Leaf class
class Developer implements Employee {
    private String name;
    private int employeeId;
    private String title;

    public Developer(String name, int employeeId, String title) {
        this.name = name;
        this.employeeId = employeeId;
        this.title = title;
    }

    @Override
    public void showDetails() {
        System.out.println("\t\tDEVELOPER ID: " + employeeId + ", NAME: " + name + ", TITLE: " + title);
    }
}


// Leaf class
class Salesperson implements Employee {
    private String name;
    private int employeeId;
    private String title;

    public Salesperson(String name, int employeeId, String title) {
        this.name = name;
        this.employeeId = employeeId;
        this.title = title;
    }

    @Override
    public void showDetails() {
        System.out.println("\t\tSALESPERSON ID: " + employeeId + ", NAME: " + name + ", TITLE: " + title);
    }
}

// Composite class
class Manager implements Employee {
    private String name;
    private int employeeId;
    private String title;
    private Set<Employee> employees;

    public Manager(String name, int employeeId, String title) {
        this.name = name;
        this.employeeId = employeeId;
        this.title = title;
        employees = new HashSet<>();
    }

    void addEmployee(Employee employee) {
        employees.add(employee);
    }

    void removeEmployee(Employee employee) {
        employees.remove(employee);
    }

    @Override
    public void showDetails() {
        System.out.println("\tMANAGER ID: " + employeeId + ", NAME: " + name + ", TITLE: " + title);
        for (var employee : employees) {
            employee.showDetails();
        }
    }
}

// Composite class
class CEO implements Employee {
    private String name;
    private int employeeId;
    private String title;
    private Set<Employee> employees;

    public CEO(String name, int employeeId, String title) {
        this.name = name;
        this.employeeId = employeeId;
        this.title = title;
        employees = new HashSet<>();
    }

    void addEmployee(Employee employee) {
        employees.add(employee);
    }

    void removeEmployee(Employee employee) {
        employees.remove(employee);
    }

    @Override
    public void showDetails() {
        System.out.println("CEO: " + employeeId + ", NAME: " + name + ", TITLE: " + title);
        for (var employee : employees) {
            employee.showDetails();
        }
    }
}
